addons/website_slides/models/slide_slide.py

```diff
@@ -315,62 +383,96 @@ def _compute_user_membership_id(self):
+                    google_drive_id = match.group(2)
+
+            slide.google_drive_id = google_drive_id
+
+    @api.onchange('url', 'document_google_url', 'image_google_url', 'video_url')
     def _on_change_url(self):
+        """ Keeping a 'onchange' because we want this behavior for the frontend.
+        Changing the document / video external URL will populate some metadata on the form view.
+        The slide metadata are also fetched in create / write overrides to ensure consistency. """
+
         self.ensure_one()
-        if self.url:
-            res = self._parse_document_url(self.url)
-            if res.get('error'):
-                raise UserError(res.get('error'))
-            values = res['values']
-            if not values.get('document_id'):
-                raise UserError(_('Please enter valid Youtube or Google Doc URL'))
-            for key, value in values.items():
-                self[key] = value
-
-    @api.onchange('binary_content')
-    def _on_change_binary_content(self):
-        """ For PDFs, we assume that it takes 5 minutes to read a page.
-            If the selected file is not a PDF, it is an image (You can
-            only upload PDF or Image file) then the slide_category is changed
-            into infographic and the uploaded binary_content is transfered to the
-            image field. (It avoids the infinite loading in PDF viewer)"""
-        if self.binary_content:
-            data = base64.b64decode(self.binary_content)
-            if data.startswith(b'%PDF-'):
-                pdf = PyPDF2.PdfFileReader(io.BytesIO(data), overwriteWarnings=False, strict=False)
-                try:
-                    pdf.getNumPages()
-                except PyPDF2.utils.PdfReadError:
-                    return
-                self.completion_time = (5 * len(pdf.pages)) / 60
-            else:
-                self.slide_category = 'infographic'
-                self.image_1920 = self.binary_content
-                self.binary_content = None
+        if self.url or self.document_google_url or self.image_google_url or self.video_url:
+            slide_metadata = self._fetch_external_metadata()
+            if slide_metadata:
+                self.update(slide_metadata)
+
+    @api.onchange('document_binary_content')
+    def _on_change_document_binary_content(self):
+        if self.slide_category == 'document' and self.source_type == 'local_file' and self.document_binary_content:
+            completion_time = self._get_completion_time_pdf(base64.b64decode(self.document_binary_content))
+            if completion_time:
+                self.completion_time = completion_time
+
+    @api.onchange('slide_category')
+    def _on_change_slide_category(self):
+        """ Prevents mis-match when ones uploads an image and then a pdf without saving the form. """
+        if self.slide_category != 'infographic' and self.image_binary_content:
+            self.image_binary_content = False
```

Comment:
As those are related, is it necessary ? Shoudln't we empty binary_content directly when leaving infographic or document category ?

Reply:
I agree there is something strange with related field and the onchange behavior.
But I don't think there is much I can do about it, except deep dive into the web framework for debugging?